File: src/main/java/com/github/novusvetus/intellijstatsplugin/CustomSaveListener.java
package com.github.novusvetus.intellijstatsplugin;

import com.intellij.openapi.editor.Document;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.fileEditor.FileDocumentManagerListener;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;

public class CustomSaveListener implements FileDocumentManagerListener {

    @Override
    public void beforeDocumentSaving(Document document) {
        FileDocumentManager instance = FileDocumentManager.getInstance();
        VirtualFile file = instance.getFile(document);
        IntelliJStatsPlugin.appendHeartbeat(file, IntelliJStatsPlugin.getProject(document), true);
    }

    @Override
    public void beforeAllDocumentsSaving() {
    }

    @Override
    public void beforeFileContentReload(@NotNull VirtualFile file, @NotNull Document document) {
    }

    @Override
    public void fileWithNoDocumentChanged(@NotNull VirtualFile file) {
    }

    @Override
    public void fileContentReloaded(@NotNull VirtualFile file, @NotNull Document document) {
    }

    @Override
    public void fileContentLoaded(@NotNull VirtualFile file, @NotNull Document document) {
    }

    @Override
    public void unsavedDocumentsDropped() {
    }
}

File: src/main/kotlin/com/github/novusvetus/intellijstatsplugin/CustomSaveListener.kt
package com.github.novusvetus.intellijstatsplugin

import com.intellij.openapi.editor.Document
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.fileEditor.FileDocumentManagerListener
import com.intellij.openapi.vfs.VirtualFile

class CustomSaveListener : FileDocumentManagerListener {
    override fun beforeDocumentSaving(document: Document) {
        val instance = FileDocumentManager.getInstance()
        val file = instance.getFile(document)
        IntelliJStatsPlugin.appendHeartbeat(file, IntelliJStatsPlugin.getProject(document), true)
    }

    override fun beforeAllDocumentsSaving() {}
    override fun beforeFileContentReload(file: VirtualFile, document: Document) {}
    override fun fileWithNoDocumentChanged(file: VirtualFile) {}
    override fun fileContentReloaded(file: VirtualFile, document: Document) {}
    override fun fileContentLoaded(file: VirtualFile, document: Document) {}
    override fun unsavedDocumentsDropped() {}
}